crates/primitives/src/transaction/mod.rs

```diff
@@ -620,6 +620,75 @@ impl TransactionSigned {
         initial_tx.hash = initial_tx.recalculate_hash();
         initial_tx
     }
+
+    /// Decodes a eip2718 transaction where tx is [id, rlp(typedtx)]
+    pub fn decode_enveloped(&self, tx: Bytes) -> Result<Self, DecodeError> {
+        let mut data = tx.as_ref();
+
+        let first_header = Header::decode(&mut data)?;
```

Comment:
this seems wrong. because the entire point of this function is that the input is likely not rlp.

Reply:
Alright, removing the first_header means this 
```
let tx_length = first_header.payload_length + first_header.length();
let hash = keccak256(&data[..tx_length]);
```
can be written as 
```
let hash = keccak256(&data[..]);
```
since we know it as LegacyTx?

Reply:
yeah, we should peek at the first byte to determine whether or not it is a type byte, or if it is a RLP header, and do the following:
 * if a type byte, parse out the first byte and treat it as an enveloped typed transaction
 * otherwise, decode the header
   * if it's a list header, it is a legacy transaction
   * if it's a string header, the input is invalid